### PR TITLE
feature/PP-204

### PR DESCRIPTION
**Pull request by franciscotobar**
## What

- Validation for smart wallet address to verify if it belongs to the EOA and its deployed

## Why

- To be able to import a smart wallet

## Refs

- [Jira issue](https://rsklabs.atlassian.net/browse/PP-204)
- https://github.com/rsksmart/rif-relay-common/pull/27
- https://github.com/rsksmart/rif-relay-client/pull/23
- https://github.com/rsksmart/rif-relay-sdk/pull/27

### CODE REVIEW

File: src/components/ActionBar.tsx [hunk 1]
```diff
@@ -1,74 +1,56 @@
...
+            console.log(
+                'Failed trying to read smart wallets, erased all previous smart wallets'
+            );
+            console.log(e);
+        }
+        console.log(tempSmartWallets);
+        const tempSmartWalletsWithBalance: SmartWalletWithBalance[] = [];
+        for (let i = 0; i < tempSmartWallets.length; i += 1) {
+            // eslint-disable-next-line no-await-in-loop
+            const tempSmartWallet = await Utils.setBalance(
```
Comment (antomor) [inline src/components/ActionBar.tsx]: I know this has been introduced in a different PR, but I'll fix it now, but calling this method in parallel.

File: src/modals/Validate.tsx [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+                state.token!
+            );
+            const tempSmartWallets = [...smartWallets, smartWalletWithBalance];
+            setSmartWallets(tempSmartWallets);
+            localStorage.setItem(
+                state.account,
+                JSON.stringify(tempSmartWallets)
+            );
+            dispatch({ type: 'set_loader', loader: false });
+            close();
```
Comment (antomor) [inline src/modals/Validate.tsx]: I think that those lines should be moved after the catch, otherwise the UI keeps waiting if there are errors

**Review by antomor — CHANGES_REQUESTED** [review]
@franciscotobar I like the new functionality, even if there are still some issues that need to be fixed.

File: src/components/ActionBar.tsx (hunk 1 above)
Comment (franciscotobar) [reply src/components/ActionBar.tsx]: Im not sure what do you mean :S

File: src/modals/Validate.tsx (hunk 2 above)
Comment (franciscotobar) [reply src/modals/Validate.tsx]: I will check it and do the necessary changes.

File: package.json [hunk 3]
```diff
@@ -3,7 +3,7 @@
   "version": "0.0.1-alpha.1",
   "private": true,
   "dependencies": {
-    "@rsksmart/rif-relay-sdk": "0.0.1-alpha.6",
+    "@rsksmart/rif-relay-sdk": "github:rsksmart/rif-relay-client#feature/PP-204",
```
Comment (antomor) [inline package.json]: we should restore this dependency once we merge the related PRs

Comment (antomor) [inline package.json]: This dependency must be restored before merging.

Comment (antomor) [reply package.json]: Furthermore, it points to rif-relay-client instead of rif-relay-sdk

File: src/Utils.ts [hunk 4]
```diff
@@ -122,6 +125,23 @@ class Utils {
         transactions.push(transaction);
         localStorage.setItem(address, JSON.stringify(transactions));
     }
+
+    static async setBalance(
```
Comment (antomor) [inline src/Utils.ts]: I think this method should be called `getSmartWalletBalance` or `getTokenBalances`

**Review by antomor — CHANGES_REQUESTED** [review]
@franciscotobar I found a small bug. If we try to click the reload button, we see the same SmartWallets multiple times in the UI.
image

File: package.json (hunk 3 above)
Comment (franciscotobar) [reply package.json]: updated

File: src/modals/TransactionHistory.tsx [hunk 5]
```diff
@@ -19,10 +19,12 @@ function TransactionHistory(props: TransactionHistoryProps) {
     const [transactions, setTransactions] = useState<Transaction[]>([]);
 
     useEffect(() => {
-        if (state.smartWallet?.address! in localStorage) {
+        if (`${state.smartWallet?.address!}-${state.chainId}` in localStorage) {
```
Comment (antomor) [inline src/modals/TransactionHistory.tsx]: I think we should have function like `getAddressKey` or similar to be reused elsewhere

File: src/modals/Validate.tsx [hunk 6]
```diff
@@ -85,7 +85,7 @@ function Validate(props: ValidateProps) {
             const tempSmartWallets = [...smartWallets, smartWalletWithBalance];
             setSmartWallets(tempSmartWallets);
             localStorage.setItem(
-                state.account,
+                `${state.account}-${state.chainId}`,
```
Comment (antomor) [inline src/modals/Validate.tsx]: I think we should have function like `getAddressKey` or similar to be reused elsewhere

File: src/modals/Validate.tsx [hunk 7]
```diff
@@ -128,7 +128,7 @@ function Validate(props: ValidateProps) {
                 ];
                 setSmartWallets(tempSmartWallets);
                 localStorage.setItem(
-                    state.account,
+                    `${state.account}-${state.chainId}`,
```
Comment (antomor) [inline src/modals/Validate.tsx]: I think we should have function like `getAddressKey` or similar to be reused elsewhere

File: src/modals/TransactionHistory.tsx (hunk 5 above)
Comment (franciscotobar) [reply src/modals/TransactionHistory.tsx]: updated

File: src/modals/Validate.tsx (hunk 6 above)
Comment (franciscotobar) [reply src/modals/Validate.tsx]: updated

File: src/modals/Validate.tsx (hunk 7 above)
Comment (franciscotobar) [reply src/modals/Validate.tsx]: updated

File: package.json [hunk 8]
```diff
@@ -3,7 +3,7 @@
   "version": "0.0.1-alpha.1",
   "private": true,
   "dependencies": {
-    "@rsksmart/rif-relay-sdk": "0.0.1-alpha.6",
+    "@rsksmart/rif-relay-sdk": "github:rsksmart/rif-relay-sdk#feature/PP-204",
```
Comment (antomor) [inline package.json]: I think we should update this now, after realising a new version of the SDK

File: src/Utils.ts [hunk 9]
```diff
@@ -120,7 +129,31 @@ class Utils {
...
+        const rbtcBalance = await Utils.getBalance(smartWallet.address);
+        const swWithBalance = {
+            ...smartWallet,
+            balance: `${Utils.fromWei(balance)} ${token.symbol}`,
+            rbtcBalance: `${Utils.fromWei(rbtcBalance)} RBTC`
+        };
+        return swWithBalance;
+    }
+
+    static getChainKey(chainId: number, address: string): string {
```
Comment (antomor) [inline src/Utils.ts]:
```suggestion
    static getTransactionKey(chainId: number, address: string): string {
```

**Review by antomor — CHANGES_REQUESTED** [review]

**Review by antomor — APPROVED** [review]
LGTM

**Pull request merged**